addon/tracked.ts

```diff
@@ -1,143 +1,180 @@
+      assert(
+        `@tracked: Stack frames do not match.\nExpected: '${frame}'\nReceived: ${lastFrame}`,
+        frame === lastFrame
+      );
     }
-    OBSERVERS_SETUP.set(this, true);
 
-    return getterProvided.call(this);
+    return value;
   };
+}
 
-  if (setterProvided) {
-    desc.set = function(value) {
-      if (typeof key === 'string') {
-        Ember.notifyPropertyChange(this, key);
-        setterProvided.call(this, value);
-      } else if (DEBUG) {
-        throw new Error(`@tracked - unsupported property type ${String(key)}`);
-      }
-    };
+function trackSet<O, K extends keyof O>(
+  key: K,
+  setter: (this: O, value: O[K]) => void = function(this: O, value: O[K]) {
+    setValue(this, key, value);
   }
+) {
+  return function(this: O, value: O[K]) {
+    setter.call(this, value);
+    invalidateDependentsOf([this, key]);
+    notifyPropertyChange(this, key);
+  };
+}
 
-  return desc;
+const DEPENDENCIES = new WeakMapWithDefault(
+  () => new MapWithDefault(() => new MapWithDefault(() => new Set<Key>()))
+);
+
+function declareDependency(from: Field<any>, on: Field<any>) {
+  DEPENDENCIES.get(on[0])
+    .get(on[1])
+    .get(from[0])
+    .add(from[1]);
 }
 
-function installTrackedProperty(_target: object, key: string | symbol, descriptor?: PropertyDescriptor & { initializer: (() => void)}): PropertyDescriptor {
-  // only happens in babel, never in TS (Sept 2018)
-  // TODO check for whether initializer is a function
-  const initializer = descriptor && descriptor.initializer;
-
-  let values = new WeakMap();
-
-  let get;
-  if (typeof initializer === 'function') {
-    get = function(this: object) {
-      if (values.has(this)) {
-        return values.get(this);
-      } else {
-        let value = initializer.call(this);
-        values.set(this, value);
-        return value;
-      }
-    };
-  } else {
-    get = function(this: object) {
-      return values.get(this);
-    }
+function getDependentsOf<O extends object>([obj, key]: Field<O>): [
+  any,
+  ReadonlySet<Key>
+][] {
+  if (DEPENDENCIES.has(obj) && DEPENDENCIES.get(obj).has(key)) {
+    return Array.from(DEPENDENCIES.get(obj).get(key));
   }
 
-  return {
-    configurable: true,
-    // TODO: correcting a misspelling caused chrome to error
-    // writable: true,
-
-    get,
-    set(value) {
-      if (typeof key === 'string') {
-        values.set(this, value);
-        Ember.notifyPropertyChange(this, key);
-      } else if (DEBUG) {
-        throw new Error(`@tracked - unsupported property type ${String(key)}`);
-      }
+  return [];
+}
+
+function invalidateDependentsOf<O extends object>(field: Field<O>) {
+  for (const [obj, keys] of getDependentsOf(field)) {
+    for (const key of keys) {
+      notifyPropertyChange(obj, key);
     }
-  };
+  }
 }
 
+function clearDependencies([obj, key]: Field<any>) {
+  // if (DEPENDENCIES.has(obj)) {
+  //   DEPENDENCIES.get(obj).delete(key);
+  // }
```

Comment:
Commented out, because this actually needs to remove the inverse of this relationship. The current data structure does not allow this efficiently.